spec/classes/stub_spec.rb

```diff
@@ -0,0 +1,275 @@
+
+      ARGV.replace []
+      allow($stdin).to receive(:read).and_return("dog\ncat\n")
+      load subject_path
+
+      ARGV.replace %w(first_argument second_argument)
+      allow($stdin).to receive(:read).and_return('')
+      load subject_path
+
+      ARGV.replace %w(first_argument second_argument)
+      allow($stdin).to receive(:read).and_return("dog\ncat\n")
+      load subject_path
+
+      YAML.load(stub_call_file.string)
+    end
+
+    it 'logs a blank STDIN for the first call' do
+      expect(stub_call_log[0]['stdin']).to be_empty
+    end
+
+    it 'logs no arguments for the first call' do
+      expect(stub_call_log[0]['args']).to be_nil
+    end
+
+    it 'logs some STDIN for the second call' do
+      expect(stub_call_log[1]['stdin']).to eql "dog\ncat\n"
+    end
+
+    it 'logs no arguments for the second call' do
+      expect(stub_call_log[1]['args']).to be_nil
+    end
+
+    it 'logs a blank STDIN for the third call' do
+      expect(stub_call_log[2]['stdin']).to be_empty
+    end
+
+    it 'logs some arguments for the third call' do
+      expect(stub_call_log[2]['args']).to eql %w(first_argument second_argument)
+    end
+
+    it 'logs some STDIN for the fourth call' do
+      expect(stub_call_log[3]['stdin']).to eql "dog\ncat\n"
+    end
+
+    it 'logs some arguments for the fourth call' do
+      expect(stub_call_log[3]['args']).to eql %w(first_argument second_argument)
+    end
+
+    it 'exits with appropriate code for first call' do
+      expect(exit_code_list[0]).to be_nil
+    end
+
+    it 'exits with appropriate code for second call' do
+      expect(exit_code_list[1]).to be_nil
+    end
+
+    it 'exits with appropriate code for third call' do
+      expect(exit_code_list[2]).to be_nil
+    end
+
+    it 'exits with appropriate code for fourth call' do
+      expect(exit_code_list[3]).to be_nil
+    end
+  end
+  context 'with configurations that does not end up matching calls' do
+    before(:each) do
+      stdout_configuration = [
+        {
+          args: ['super_special_argument'],
+          outputs: [
+            {
+              target: :stderr,
+              content: "no args content\n"
+            }
+          ],
+          statuscode: 1
+        }
+      ]
+      allow($stdin).to receive(:read).and_return('')
+      allow(stub_config_pathname).to receive(:exist?).and_return(true)
+      allow(stub_config_pathname).to receive(:read).and_return(stdout_configuration.to_yaml)
+
+      ARGV.replace []
+      begin
+        load subject_path
+      rescue SystemExit
+      end
+
+      ARGV.replace ['wrongo']
+      begin
+        load subject_path
+      rescue SystemExit
+      end
+    end
+
+    it 'exits with appropriate code for first call' do
+      expect(exit_code_list[0]).to eql 0
+    end
+
+    it 'exits with appropriate code for second call' do
+      expect(exit_code_list[1]).to eql 0
+    end
+  end
+  context 'with configurations for stdout, stderr and file behaviour' do
```

Comment:
This particular context is included last, because the coverage generator doesn't deal well with re-loading a file over and over again (it basically clobbers the coverage). With this as the final run, we do get full coverage (or as full as simplecov can give).